Pendulum/Data/CDPenPal+Extensions.swift

```diff
@@ -0,0 +1,270 @@
+                for item in unusedStationery {
+                    if !setOfResults.contains(item) {
+                        intermediate.append(ParameterCount(name: item, count: 0))
+                    }
+                }
+            }
+            return intermediate
+        } catch {
+            dataLogger.error("Could not fetch distinct stationery: \(error.localizedDescription)")
+        }
+        return []
+    }
+    
+    func fetchDistinctStationery(ofType stationery: String) -> [ParameterCount] {
+        CDPenPal.fetchDistinctStationery(ofType: stationery, for: self)
+    }
+    
+    static func fetch(withStatus eventType: EventType? = nil) -> [CDPenPal] {
+        let fetchRequest = NSFetchRequest<CDPenPal>(entityName: CDPenPal.entityName)
+        if let eventType = eventType {
+            fetchRequest.predicate = NSPredicate(format: "lastEventTypeValue = %d", eventType.rawValue)
+        }
+        do {
+            return try PersistenceController.shared.container.viewContext.fetch(fetchRequest)
+        } catch {
+            dataLogger.error("Could not fetch penpals with status \(eventType?.description ?? "all"): \(error.localizedDescription)")
+        }
+        return []
+    }
+    
+    func archive(_ value: Bool = true) {
+        self.archived = value
+        PersistenceController.shared.save()
+    }
+    
+    func update(from contact: CNContact) {
+        self.image = contact.thumbnailImageData
+        self.initials = contact.initials
+        self.name = contact.fullName ?? self.name
+        PersistenceController.shared.save()
+    }
+    
+    func delete() {
+        PersistenceController.shared.container.viewContext.delete(self)
+        PersistenceController.shared.save()
+    }
+    
+    static func syncWithContacts() async {
+
+        let store = CNContactStore()
+        let keys = [
+            CNContactFormatter.descriptorForRequiredKeys(for: .fullName),
+            CNContactOrganizationNameKey,
+            CNContactImageDataAvailableKey,
+            CNContactThumbnailImageDataKey
+        ] as! [CNKeyDescriptor]
+        
+        let fetchRequest = NSFetchRequest<CDPenPal>(entityName: CDPenPal.entityName)
+        let penpals = (try? PersistenceController.shared.container.viewContext.fetch(fetchRequest)) ?? []
+        let mapping = UserDefaults.shared.penpalContactMap
+        
+        for penpal in penpals {
+            guard let uuid = penpal.id, let contactID = mapping[uuid.uuidString] else { continue }
+            do {
+                let contact = try store.unifiedContact(withIdentifier: contactID, keysToFetch: keys)
+                penpal.update(from: contact)
+            } catch {
+                appLogger.error("Could not fetch contact with ID \(contactID) \(penpal.wrappedName): \(error.localizedDescription)")
+            }
+        }
+        
+        let penpalsWithNoContact: [String: [CDPenPal]] = Dictionary(grouping: penpals.filter {
```

Comment:
Did we verify if they are consistent over devices?

Reply:
They are not. Which is why I've changed it so Contact ID no longer syncs, is stored locally in UserDefaults instead, and when a new contact comes in that isn't associated with a local contact, the app tries to find a matching one and join them up.